index.html

```diff
@@ -3310,6 +3346,11 @@ <h3>Processing Errors</h3>
 in the `proofPurpose` property in the proof. See Section
 <a href="#retrieve-verification-method"></a>.
           </dd>
+          <dt id="PROOF_CHAIN_CYCLE_ERROR">PROOF_CHAIN_CYCLE_ERROR (-26)</dt>
+          <dd>
+A proof cycle was detected during the verification of a proof set or proof
+chain. See Section <a href="#verify-proof-sets-and-chains"></a>.
+          </dd>
```

Comment:
I didn't see that we're already iterating over `allProofs` when doing proof chain checking. I was expecting that we would be picking out the proof we wanted to check as the starting point instead. This eliminates the need for my other comments above -- and, I believe, eliminates the need for this error as well. I don't think the cycles can happen without a proof failing to verify given the other changes we've made.

However, I do think we should modify this so that we don't have to process `allProofs`, as it places the burden on the presenter to avoid including any proofs that might not be accepted by the verifier, but that would not stop the verifier from being able to verify a particular proof (and its dependencies) that it does accept.

I think we need to consider how to do this, perhaps by keeping track of the dependencies as we go and allowing for the proof status to be verified / not verified / not considered (a third value). The "not considered" value would be perfectly fine if the proof was unrelated to the target proof that is being considered for verification.

Reply:
Hi Dave (@dlongley), removed the PROOF_CHAIN_CYCLE  error.

As for your other comment. Currently we say: "For each proof in allProofs, do the following steps: ". Couldn't we just modify this one line to something like

"For each proof in the secured document to be verified, do the following steps:"

That way no one is forced to verify all the proofs and the algorithm still works as it stands.

Reply:
@Wind4Greg,

> "For each proof in the secured document to be verified, do the following steps:"...

Well, I think this implies that the verifier knows those proofs to begin with when I would expect them to only know that there's a particular proof they want to verify ... which itself allows the verifier to discover what others need verification (through `previousProof`).

We might need to have a previous step that will walk through `previousProof` recursively and build a list of proofs to verify -- stopping recursion whenever we see a proof we've already seen. We don't need to throw a cycle error this way and we won't get stuck in a loop and we will only return that verification passed if each of those proofs verifies.

So I think we need some (ideally short) text that covers this simply enough that can be put as a starting step to gather all the proofs that need verification (vs. just assuming the verifier knows what they will be) and then we can make the statement you're saying without issue, I'd think. Then we'd be good here, IMO.

Reply:
@dlongley This is starting to seem like an optimization. I don't have any examples of elaborate proof sets/chains with chains that don't depend on some of the proofs to test against. If you have a suggested algorithm change let me know. I'm leaning on putting some text at the beginning to let them know that they can optimize of just produce the recursively generated list at the beginning.

Reply:
@Wind4Greg,

> I'm leaning on putting some text at the beginning to let them know that they can optimize of just produce the recursively generated list at the beginning.

I'd say go ahead and add some text to this effect and we'll see what reviewers think. It's not core to the algorithm and the `previousProof` indicates how you can find out which proofs are related -- verification implementers can process this however they like to obtain the proofs that need to be verified, but they must do so to ensure that the whole chain is properly verified.

Reply:
A test that will fail verification if one of the (deep) dependency proofs is left out would be a good way to ensure implementers are doing this this check.

Reply:
@dlongley I just added intro text to the chain verification algorithm stating that the algorithm given is one approach and that others are possible, particularly if only subset of the `allProofs` needs to be verified.